core/common/src/main/java/alluxio/underfs/NonAtomicCreateUnderFileSystem.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.AlluxioURI;
+import alluxio.Constants;
+import alluxio.exception.ExceptionMessage;
+import alluxio.security.authorization.Permission;
+import alluxio.underfs.options.CreateOptions;
+import alluxio.underfs.options.NonAtomicCreateOptions;
+import alluxio.util.IdUtils;
+import alluxio.util.io.PathUtils;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * An {@link UnderFileSystem} which does not have an atomic create operation.
+ */
+public abstract class NonAtomicCreateUnderFileSystem extends UnderFileSystem {
```

Comment:
Is there a way to just have the default implementation of `UFS.create()` assume non-atomic creates, but the object stores can override it to do its own atomic creates? This allows new UFS implementations to not have to worry about atomicity, but later can optimize it if possible.

What do you think?